crypto/http/http_client.c

```diff
@@ -0,0 +1,1238 @@
+        return 0;
+    elapsed_time = (long)(time(NULL) - start_time); /* this might overflow */
+    return timeout <= elapsed_time ? -1 : timeout - elapsed_time;
+}
+
+/*-
+ * Exchange HTTP request and response with the given server.
+ * If req_mem == NULL then use GET and ignore content_type, else POST.
+ * The redirection_url output (freed by caller) parameter is used only for GET.
+ *
+ * Typically the wbio and rbio parameters are NULL and a network BIO is created
+ * internally for connecting to the given server and port, optionally via a
+ * proxy and its port, and is then used for exchanging the request and response.
+ * If wbio is given and rbio is NULL then this BIO is used instead.
+ * If both wbio and rbio are given (which may be memory BIOs for instance)
+ * then no explicit connection is attempted,
+ * wbio is used for writing the request, and rbio for reading the response.
+ *
+ * bio_update_fn is an optional BIO connect/disconnect callback function,
+ * which has the prototype
+ *   BIO *(*OSSL_HTTP_bio_cb_t) (BIO *bio, void *arg, int conn, int detail);
+ * The callback may modify the HTTP BIO provided in the bio argument,
+ * whereby it may make use of any custom defined argument 'arg'.
+ * During connection establishment, just after BIO_connect_retry(),
+ * the callback function is invoked with the 'conn' argument being 1
+ * 'detail' indicating whether a HTTPS (i.e., TLS) connection is requested.
+ * On disconnect 'conn' is 0 and 'detail' indicates that no error occurred.
+ * For instance, on connect the funct may prepend a TLS BIO to implement HTTPS;
+ * after disconnect it may do some error diagnostics and/or specific cleanup.
+ * The function should return NULL to indicate failure.
+ * After disconnect the modified BIO will be deallocated using BIO_free_all().
+ */
+BIO *OSSL_HTTP_transfer(const char *server, const char *port, const char *path,
+                        int use_ssl, const char *proxy, const char *proxy_port,
+                        BIO *wbio, BIO *rbio,
+                        OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
+                        const STACK_OF(CONF_VALUE) *headers,
+                        const char *content_type, BIO *req_mem,
+                        int maxline, unsigned long max_resp_len, int timeout,
+                        const char *expected_ct, int expect_asn1,
+                        char **redirection_url)
+{
+    time_t start_time = timeout > 0 ? time(NULL) : 0;
+    BIO *cbio; /* = wbio if present, used as connection BIO if rbio is NULL */
+    OSSL_HTTP_REQ_CTX *rctx;
+    BIO *resp = NULL;
+
+    if (redirection_url != NULL)
+        *redirection_url = NULL; /* do this beforehand to prevent dbl free */
+
+    if (use_ssl && bio_update_fn == NULL) {
+        HTTPerr(0, HTTP_R_TLS_NOT_ENABLED);
+        return NULL;
+    }
+    if (rbio != NULL && (wbio == NULL || bio_update_fn != NULL)) {
+        HTTPerr(0, ERR_R_PASSED_INVALID_ARGUMENT);
+        return NULL;
+    }
```

Comment:
What about the case where `wbio != NULL` and `rbio == NULL`?

Reply:
As (newly ) documented, this selects the case that `wbio` is used for connecting, writing, and reading. 
I might rename the `wbio` parameter to `bio` (or `cwrbio`) - what you think?

Reply:
I've changed the `wbio` parameter name to `bio` because it may be used more generally.